Test failure due to correlated subquery in IN predicate

TestRedisDistributedHash.testCorrelatedInPredicateSubqueries fails with the following error:

```
java.lang.AssertionError: Execution of 'actual' query failed: SELECT orderkey, clerk IN (SELECT clerk FROM orders s WHERE s.custkey = o.custkey AND s.orderkey < o.orderkey) FROM orders o

	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:86)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:117)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:112)
	at com.facebook.presto.tests.AbstractTestQueries.testCorrelatedInPredicateSubqueries(AbstractTestQueries.java:7166)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:127)
Caused by: java.lang.IllegalStateException: No mapping for column handle RedisColumnHandle{connectorId=redis, ordinalPosition=0, name=redis_key, type=varchar(64), mapping=null, dataFormat=null, formatHint=null, keyDecoder=true, hidden=true, internal=false}!
	at com.google.common.base.Preconditions.checkState(Preconditions.java:518)
	at com.facebook.presto.redis.decoder.hash.HashRedisRowDecoder.decodeRow(HashRedisRowDecoder.java:58)
	at com.facebook.presto.redis.RedisRecordCursor.nextRow(RedisRecordCursor.java:184)
	at com.facebook.presto.redis.RedisRecordCursor.advanceNextPosition(RedisRecordCursor.java:147)
	at com.facebook.presto.spi.RecordPageSource.getNextPage(RecordPageSource.java:99)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:262)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:303)
	at com.facebook.presto.operator.Driver.lambda$processFor$6(Driver.java:234)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:537)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:229)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:623)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:463)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

This seems to be a regression introduced in 98faf69c110b0f7e51b79924fa06aef9ae345a6e

--- a/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
+++ b/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java
@@ -167,10 +167,7 @@ public class PlanOptimizers
                 new UnaliasSymbolReferences(),
                 new IterativeOptimizer(
                         stats,
-                        ImmutableSet.of(
-                                new RemoveRedundantIdentityProjections(),
-                                new TransformCorrelatedInPredicateToJoin(),
-                                new ImplementFilteredAggregations())
+                        ImmutableSet.of(new RemoveRedundantIdentityProjections())
                 ),
                 new SetFlatteningOptimizer(),
                 new ImplementIntersectAndExceptAsUnion(),
@@ -188,6 +185,12 @@ public class PlanOptimizers
                         stats,
                         ImmutableList.of(new TransformCorrelatedScalarAggregationToJoin(metadata.getFunctionRegistry())),
                         ImmutableSet.of(new com.facebook.presto.sql.planner.iterative.rule.TransformCorrelatedScalarAggregationToJoin(metadata.getFunctionRegistry()))),
+                new IterativeOptimizer(
+                        stats,
+                        ImmutableSet.of(
+                                new TransformCorrelatedInPredicateToJoin(), // must be run after PruneUnreferencedOutputs
+                                new ImplementFilteredAggregations())
+                ),
                 new PredicatePushDown(metadata, sqlParser),
                 new PruneUnreferencedOutputs(),
                 new IterativeOptimizer(

--- a/presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java
+++ b/presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java
@@ -49,6 +49,7 @@ import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.join;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.node;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.semiJoin;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.strictTableScan;
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.tableScan;
 import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
 import static com.facebook.presto.sql.planner.optimizations.Predicates.isInstanceOfAny;
@@ -295,6 +296,27 @@ public class TestLogicalPlanner
                                                 tableScan("orders", ImmutableMap.of("orderkey", "orderkey"))
                                         )
                                 )
+                        )
+                )
+        );
+    }
+
+    /**
+     * Handling of correlated in predicate involves group by over all symbols from source. Once aggregation is added to the plan,
+     * it prevents pruning of the unreferenced symbols. However, the aggregation's result doesn't actually depended on those symbols
+     * and this test makes sure the symbols are pruned first.
+     */
+    @Test
+    public void testSymbolsPrunedInCorrelatedInPredicateSource()
+    {
+        assertPlan(
+                "SELECT orderkey, comment IN (SELECT clerk FROM orders s WHERE s.orderkey = o.orderkey AND s.orderkey < 7) FROM lineitem o",
+                anyTree(
+                        node(JoinNode.class,
+                                anyTree(strictTableScan("lineitem", ImmutableMap.of(
+                                        "orderkey", "orderkey",
+                                        "comment", "comment"))),
+                                anyTree(tableScan("orders"))
                         )
                 )
         );

